packages/propeller/src/StreamsSink.mts

```diff
@@ -16,38 +16,116 @@ function groupByStream(items: [StreamName, ITimelineEvent][]): Map<StreamName, I
 
 const tracer = trace.getTracer("@equinox-js/propeller")
 
+const createLinkedAbortController = (signal: AbortSignal) => {
+  const ctrl = new AbortController()
+  signal.addEventListener("abort", () => ctrl.abort())
+  return ctrl
+}
+
+class Stream {
+  events: ITimelineEvent[] = []
+  constructor(public name: StreamName) {}
+
+  merge(event: ITimelineEvent): void {
+    const last = this.events[this.events.length - 1]
+    if (!last || last.index === event.index - 1n) {
+      this.events.push(event)
+      return
+    }
+    if (last.index > event.index) return
+    if (event.index > last.index) throw new Error("Unexpected gap")
+  }
+}
+
+const getOrAdd = <K, V>(map: Map<K, V>, key: K, val: () => V) => {
+  const current = map.get(key)
+  if (current == null) {
+    const value = val()
+    map.set(key, value)
+    return value
+  }
+  return current
+}
+
 export class StreamsSink implements Sink {
-  private readonly limit: LimitFunction
+  private queue = new Queue<Stream>()
+  private streams = new Map<StreamName, Stream>()
+  private batchStreams = new Map<IngesterBatch, Set<StreamName>>()
```

Comment:
these ultimately need to become pairs of stream+position

you want to be able to have a handler respond to a dispatch of N events on a stream by saying it did some, none or all of them (or even set a position beyond where reading was)

in a case where there is a dominant long stream that takes a long time to drain and/or never gets drained, you need to be able to keep declaring earlier batches closed per what the handler has achieved in order to unlock read ahead capacity

and/or allow stuff that's been read be dumped if the handler determines a consumer is way ahead

this becomes particularly relevant if there's streams that take inordinately long to process relative to others (could be rate limiting, could be some wierd data distribution)